File: oceanfft/components/Ocean.cs
using System;
using Godot;
using Godot.Collections;

[Tool]
public partial class Ocean : Resource {
    // ## Handles updating the displacement/normal maps for the water material as well as
    // ## managing wave generation pipelines.
    [Export]
    private ShaderMaterial waterMat;
    [Export]
    private ShaderMaterial sprayMat;
    // ## The parameters for wave cascades. Each parameter set represents one cascade.
    // ## Recreates all compute piplines whenever a cascade is added or removed!
    [Export]
    Array<WaveCascadeParameters> Parameters {
        set {
            var newSize = value.Count;
            RandomNumberGenerator rng = new();
            // # All below logic is basically just required for using in the editor!
            for (int i = 0; i < newSize; i++){
                // # Ensure all values in the array have an associated cascade
                if (value[i] != null)
                {
                    value[i] = new WaveCascadeParameters();
                }
                value[i].scaleChanged = UpdateScalesUniform;
                value[i].spectrumSeed = new(rng.RandiRange(-10000, 10000), rng.RandiRange(-10000, 10000));
                value[i].time = 120.0f + Mathf.Pi * i; // # We make sure to choose a time offset such that cascades don't interfere!
            }
            parameters = value;
            SetupWaveGenerator();
            UpdateScalesUniform();
        }
        get {
            return parameters;
        }
    }
    Array<WaveCascadeParameters> parameters;
    [ExportGroup("Performance Parameters")]
    [Export(PropertyHint.Enum, "128x128:128,256x256:256,512x512:512,1024x102:1024")] int MapSize {
        set {
            mapSize = value;
            SetupWaveGenerator();
        }
    }
    int mapSize = 1024;
    // ## How many times the wave simulation should update per second.
    // ## Note: This doesn't reduce the frame stutter caused by FFT calculation, only
    // ##       minimizes GPU time taken by it!
    [Export(PropertyHint.Range, "0,120")] float UpdatesPerSecond {
        set {
            if(value > 0){
                updateTime = 1f / value;
                updatesPerSecond = value;
            } else {
                updateTime = 0f;
            }
        }
    }
    float updatesPerSecond = 50.0f;
    float updateTime = 0.15f;
    WaveGenerator waveGenerator {
        
    }



}


var wave_generator : WaveGenerator :
	set(value):
		if wave_generator: wave_generator.queue_free()
		wave_generator = value
		add_child(wave_generator)
var rng = RandomNumberGenerator.new()
var time := 0.0
var next_update_time := 0.0

var displacement_maps := Texture2DArrayRD.new()
var normal_maps := Texture2DArrayRD.new()

func _init() -> void:
	rng.set_seed(1234) # This seed gives big waves!

func _process(delta : float) -> void:
	# Update waves once every 1.0/updates_per_second.
	_update_water(delta)

	# if updates_per_second == 0 or time >= next_update_time:
	# 	var target_update_delta := 1.0 / (updates_per_second + 1e-10)
	# 	var update_delta := delta if updates_per_second == 0 else target_update_delta + (time - next_update_time)
	# 	next_update_time = time + target_update_delta
	# 	_update_water(update_delta)
	time += delta

func _setup_wave_generator() -> void:
	if parameters.size() <= 0: return
	for param in parameters:
		param.should_generate_spectrum = true

	wave_generator = WaveGenerator.new()
	wave_generator.map_size = map_size
	wave_generator.init_gpu(maxi(2, parameters.size())) # FIXME: This is needed because my RenderContext API sucks...

	displacement_maps.texture_rd_rid = RID()
	normal_maps.texture_rd_rid = RID()
	displacement_maps.texture_rd_rid = wave_generator.descriptors[&'displacement_map'].rid
	normal_maps.texture_rd_rid = wave_generator.descriptors[&'normal_map'].rid

	RenderingServer.global_shader_parameter_set(&'num_cascades', parameters.size())
	RenderingServer.global_shader_parameter_set(&'displacements', displacement_maps)
	RenderingServer.global_shader_parameter_set(&'normals', normal_maps)

func _update_scales_uniform() -> void:
	var map_scales : PackedVector4Array; map_scales.resize(len(parameters))
	for i in len(parameters):
		var params := parameters[i]
		var uv_scale := Vector2.ONE / params.tile_length
		map_scales[i] = Vector4(uv_scale.x, uv_scale.y, params.displacement_scale, params.normal_scale)
	# No global shader parameter for arrays :(
	WATER_MAT.set_shader_parameter(&'map_scales', map_scales)
	SPRAY_MAT.set_shader_parameter(&'map_scales', map_scales)

func _update_water(delta : float) -> void:
	if wave_generator == null: _setup_wave_generator()
	wave_generator.update(delta, parameters)

func _notification(what: int) -> void:
	if what == NOTIFICATION_PREDELETE:
		displacement_maps.texture_rd_rid = RID()
		normal_maps.texture_rd_rid = RID()

File: src/ocean.rs
use core::f32;
use std::io::Error;

use godot::obj::WithBaseField;
use godot::prelude::*;
use godot::classes::{RandomNumberGenerator, RenderingServer, Resource, ShaderMaterial, Texture2DArrayRd, Time};
use crate::wave_cascade_parameters::WaveCascadeParameters;
use crate::wave_generator::{WaveGenerator, DESCRIPTOR};
#[derive(GodotClass)]
#[class(tool, base=Node)]

struct Ocean {
    #[export]
    water_material: Option<Gd<ShaderMaterial>>,
    #[export]
    spray_material: Option<Gd<ShaderMaterial>>,
    #[export]
    #[var(get = get_parameters, set = set_parameters)]
    parameters: Array<Option<Gd<WaveCascadeParameters>>>,
    #[export(enum = (_128x128 = 128, _256x256 = 256, _512x512 = 512, _1024x1024 = 1024))]
    #[var(set = set_map_size, get = get_map_size)]
    map_size: i32,
    #[export(range = (0.0, 120.0, 1.0))]
    #[var(set = set_updates_per_second)]
    updates_per_second: real,
    update_time: real,
    #[export]
    #[var(set = set_wave_generator)]
    wave_generator: Option<Gd<WaveGenerator>>,
    rng: Gd<RandomNumberGenerator>,
    #[var]
    internal_time: f64,
    #[var]
    external_time: f64,
    #[export]
    sync_time: f32, // Minimum difference between external and internal time to sync the two
    time: f32,
    displacement_maps: Gd<Texture2DArrayRd>,
    normal_maps: Gd<Texture2DArrayRd>,

    initialized: bool,
    base: Base<Node>
}
#[godot_api]
impl INode for Ocean {
    fn get_configuration_warnings(&self) -> PackedStringArray {
        let mut s = PackedStringArray::new();
        if self.parameters.len() == 0 {
            s.push("No paremeters set");
        } else {
            for i in 0..self.parameters.len() {
                if self.parameters.at(i) == None {
                    s.push(format!("Parameter {i} is null").as_str());
                }
            }
        }
        return s;
    }
    fn init(base: Base<Node>) -> Self {
        let mut rng = RandomNumberGenerator::new_gd();
        // rng.set_seed(Time::singleton().get_unix_time_from_system().round() as u64); // Set the seed randomly upon initialize
        Ocean {
            water_material:None,
            spray_material:None,
            parameters:Array::new(),
            map_size:1024,
            updates_per_second:60.0,
            update_time:1.0/60.0,
            wave_generator:None,
            rng:rng,
            internal_time:0.0,
            external_time:0.0,
            sync_time: 0.2, 
            time: 0.0,
            displacement_maps:Texture2DArrayRd::new_gd(),
            normal_maps:Texture2DArrayRd::new_gd(),
            initialized:false,
            base,
        }
    }
    fn process(&mut self, delta: f64){
        // Update the aves every update_time
        self.time += delta as f32;
        self.internal_time += delta;
        self.external_time += delta;
        if (self.external_time - self.internal_time).abs() > self.sync_time as f64 {
            self.time += (self.external_time - self.internal_time) as f32;
            self.internal_time = self.external_time;
        }
        if self.updates_per_second == 0.0 || self.time > self.update_time {
            self.time -= self.update_time;
            godot_print!("Starting water update");
            self._update_water(self.time as f64);

        }
    }
}
#[godot_api]
impl Ocean {
    #[func]
    pub fn initialize_random(&mut self){
        self.rng.set_seed(Time::singleton().get_unix_time_from_system().round() as u64);
        self.initialized = true;
    }
    #[func]
    pub fn initialize_set(&mut self, seed: u64){
        self.rng.set_seed(seed);
        self.initialized = true;
    }
    #[func]
    pub fn set_wave_generator(&mut self, gen: Option<Gd<WaveGenerator>>){
        if self.wave_generator != None {
            self.wave_generator.as_mut().unwrap().queue_free();
        }
        self.wave_generator = gen;
        if self.wave_generator != None {
            let gen = self.wave_generator.clone();
            let mut mutself = self.base_mut();
            mutself.add_child(&gen.clone().unwrap());
        }
    }
    #[func]
    pub fn set_updates_per_second(&mut self, value: real){
        self.updates_per_second = value;
        self.update_time = 1.0 / value;
    }
    #[func]
    pub fn set_map_size(&mut self, value : i32) {
        self.map_size = value;
        self.setup_wave_generator();
    }
    #[func]
    pub fn get_map_size(&self) -> i32 {
        return self.map_size;
    }
    #[func]
    pub fn get_parameters(&self) -> Array<Option<Gd<WaveCascadeParameters>>>{
        return self.parameters.clone();
    }
    #[func]
    pub fn set_parameters(&mut self, mut val : Array<Option<Gd<WaveCascadeParameters>>>){
        let newSize = val.len();
        let mut rng = RandomNumberGenerator::new_gd();
        for i in 0..newSize {
            match val.at(i) {
                Some(mut x) => {
                    let mut param = x.bind_mut();
                    param.spectrum_seed = Vector2i { x: rng.randi_range(-10000, 10000), y: rng.randi_range(-10000, 10000) };
                    param.time = 120.0 + f32::consts::PI * (i as f32);
                    param.signals().scale_changed().connect_other(self, Ocean::scale_changed);
                }
                None => {
                    // val.set(i, WaveCascadeParameters::new_gd()); // Not sure how to set this up. That's for future me to figure out ig
                }
                
            }
        }
        self.parameters = val;
        self.setup_wave_generator();
        self.update_scales_uniform();
    }
    fn _update_water(&mut self, time: f64){
        if self.wave_generator == None {
            godot_print!("Setting up wave generator");
            self.setup_wave_generator();
            return; // Returning so that if we get a failed setup, we don't have issues
        }
        godot_print!("Wave generator is set up, updating");
        self.wave_generator.as_mut().unwrap().bind_mut().update(time, self.parameters.clone());
    }
    pub fn scale_changed(&mut self){

    }
    fn setup_wave_generator(&mut self){
        if self.parameters.len() == 0 {
            return;
        }
        godot_print!("Creating new waveGen");
        let mut wave_gen_gd = WaveGenerator::new_alloc();
        godot_print!("Wavegen created");
        let do_steps = || -> Result<(), Error> {
            {
                let mut wave_gen = wave_gen_gd.bind_mut();
                wave_gen.map_size = self.map_size;
                wave_gen.init_gpu(2.max(self.parameters.len() as u32));
                self.displacement_maps.set_texture_rd_rid(wave_gen.descriptors[DESCRIPTOR::DisplacementMap as usize].rid);
                self.normal_maps.set_texture_rd_rid(wave_gen.descriptors[DESCRIPTOR::NormalMap as usize].rid);
                RenderingServer::singleton().global_shader_parameter_set("num_cascades", &(self.parameters.len() as u32).to_variant());
                RenderingServer::singleton().global_shader_parameter_set("displacements", &self.displacement_maps.to_variant());
                RenderingServer::singleton().global_shader_parameter_set("normals", &self.normal_maps.to_variant());
            }
            self.wave_generator = Some(wave_gen_gd.clone());
            Ok(())
        }();
        match do_steps {
            Ok(x) => {
                // godot_print!("Steps done successfully");
            }
            Err(e) => {
                godot_print!("Failed from do_steps");
                wave_gen_gd.free();
                godot_error!("ocean.rs, line 184\n{}", e);
            }

        }
    }
    fn update_scales_uniform(&mut self){
        let mut map_scales: PackedVector4Array = PackedVector4Array::new();
        map_scales.resize(self.parameters.len());
        for i in 0..self.parameters.len() {
            let param = self.parameters.at(i).unwrap();
            let uv_scale = Vector2::ONE / param.bind().get_tile_length();
            
            map_scales[i] = Vector4 { x: uv_scale.x, y: uv_scale.y, z: param.bind().get_displacement_scale(), w: param.bind().get_normal_scale() };
        }
        if self.water_material == None || self.spray_material == None {
            return;
        }
        self.water_material.as_mut().unwrap().set_shader_parameter("map_scales", &map_scales.to_variant());
        self.spray_material.as_mut().unwrap().set_shader_parameter("map_scales", &map_scales.to_variant());
    }
}